python/kfserving/test/test_storage.py

```diff
@@ -94,61 +96,62 @@ def test_https_uri_path(_):
-
-
-@mock.patch('requests.get', return_value=MockHttpResponse(status_code=200, content_type='application/x-tar',
-                                                          raw=FILE_TAR_GZ_RAW))
-def test_http_uri_path_with_tar_gz_query_params(_):
-    with tempfile.TemporaryDirectory() as out_dir:
-        https_with_query_uri = 'https://foo.bar/model.tar.gz?foo=bar'
-        assert kfserving.Storage.download(https_with_query_uri, out_dir=out_dir) == out_dir
-        assert os.path.exists(os.path.join(out_dir, 'model.pth'))
-
-
-@mock.patch('requests.get', return_value=MockHttpResponse(status_code=200, content_type='application/zip',
-                                                          raw=FILE_ZIP_RAW))
-def test_http_uri_path_with_zip(_):
-    with tempfile.TemporaryDirectory() as out_dir:
-        https_uri = 'https://foo.bar/model.zip'
-        assert kfserving.Storage.download(https_uri, out_dir=out_dir) == out_dir
-        assert os.path.exists(os.path.join(out_dir, 'model.pth'))
-
-
-@mock.patch('requests.get', return_value=MockHttpResponse(status_code=200, content_type='application/zip',
-                                                          raw=FILE_ZIP_RAW))
-def test_http_uri_path_with_zip_query_params(_):
-    with tempfile.TemporaryDirectory() as out_dir:
-        https_with_query_uri = 'https://foo.bar/model.zip?foo=bar'
-        assert kfserving.Storage.download(https_with_query_uri, out_dir=out_dir) == out_dir
-        assert os.path.exists(os.path.join(out_dir, 'model.pth'))
-
-
-@mock.patch('requests.get', return_value=MockHttpResponse(status_code=404))
-def test_nonexistent_uri(_):
-    non_existent_uri = 'https://theabyss.net/model.joblib'
-    with pytest.raises(RuntimeError):
-        kfserving.Storage.download(non_existent_uri)
-
-
-@mock.patch('requests.get', return_value=MockHttpResponse(status_code=200))
-def test_uri_no_filename(_):
-    bad_uri = 'https://foo.bar/test/'
-    with pytest.raises(ValueError):
-        kfserving.Storage.download(bad_uri)
-
-
-@mock.patch('requests.get', return_value=MockHttpResponse(status_code=200, content_type='text/html'))
-def test_html_content_type(_):
-    bad_uri = 'https://some.site.com/test.model'
-    with pytest.raises(RuntimeError):
-        kfserving.Storage.download(bad_uri)
+def gen_download_tests(uri, response, expected_error=None):
+    @mock.patch('requests.get', return_value=response)
+    def test_success(_):
+        with tempfile.TemporaryDirectory() as out_dir:
+            assert kfserving.Storage.download(uri, out_dir=out_dir) == out_dir
+            assert os.path.exists(os.path.join(out_dir, 'model.pth'))
+
+    @mock.patch('requests.get', return_value=response)
+    def test_fail(_):
+        with pytest.raises(expected_error):
+            kfserving.Storage.download(uri)
+
+    if expected_error is None:
+        return test_success
+    return test_fail
+
+
+test_http_uri_path_with_xtar = gen_download_tests(HTTPS_URI_TARGZ,
```

Comment:
I think we can use `pytest.mark.parametrize` here for different combinations of HTTP_URI and HttpResponse ?